### PR TITLE
Implements creating a term suggester from x-content

### PR DESCRIPTION
**Pull request by abeyad**

### CODE REVIEW

**Comment by cbuescher** [comment]
@abeyad I took a first look and only left a few minor things as comments. Can you also activate the testFromXContent() tests in TermSuggestionBuilderTests? When I did that they were still failing for me. Maybe you can take a look, I guess there might be still some things missing from the parser or some parameters not surviving the roundtrip correctly.

**Comment by abeyad** [comment]
@cbuescher I implemented the changes and all tests pass now.

File: core/src/main/java/org/elasticsearch/search/suggest/term/TermSuggestionBuilder.java [hunk 1]
```diff
@@ -457,9 +490,9 @@ public SortBy readFrom(final StreamInput in) throws IOException {
             return values()[ordinal];
         }
 
-        public static SortBy fromString(final String str) {
+        public static SortBy resolve(final String str) {
             Objects.requireNonNull(str, "Input string is null");
-            return valueOf(str.toUpperCase(Locale.ROOT));
+            return valueOf(str.toUpperCase(Locale.US));
```
Comment (cbuescher) [inline core/src/main/java/org/elasticsearch/search/suggest/term/TermSuggestionBuilder.java]: nit: we almost never use `Locale.US` exept for some number formating. While I think it doesn't make a difference for the enum names in question here, I'd suggest going with `Locale.ROOT`

File: core/src/main/java/org/elasticsearch/search/suggest/term/TermSuggestionBuilder.java [hunk 2]
```diff
@@ -493,9 +526,18 @@ public StringDistanceImpl readFrom(final StreamInput in) throws IOException {
             return values()[ordinal];
         }
 
-        public static StringDistanceImpl fromString(final String str) {
+        public static StringDistanceImpl resolve(final String str) {
             Objects.requireNonNull(str, "Input string is null");
-            return valueOf(str.toUpperCase(Locale.ROOT));
+            try {
+                return valueOf(str.toUpperCase(Locale.US));
```
Comment (cbuescher) [inline core/src/main/java/org/elasticsearch/search/suggest/term/TermSuggestionBuilder.java]: same here

File: core/src/main/java/org/elasticsearch/search/suggest/term/TermSuggestionBuilder.java [hunk 3]
```diff
@@ -428,9 +461,9 @@ public SuggestMode readFrom(final StreamInput in) throws IOException {
             return values()[ordinal];
         }
 
-        public static SuggestMode fromString(final String str) {
+        public static SuggestMode resolve(final String str) {
             Objects.requireNonNull(str, "Input string is null");
-            return valueOf(str.toUpperCase(Locale.ROOT));
+            return valueOf(str.toUpperCase(Locale.US));
```
Comment (cbuescher) [inline core/src/main/java/org/elasticsearch/search/suggest/term/TermSuggestionBuilder.java]: Also please `Locale.ROOT`

File: core/src/main/java/org/elasticsearch/search/suggest/term/TermSuggestionBuilder.java (hunk 2 above)
Comment (abeyad) [inline core/src/main/java/org/elasticsearch/search/suggest/term/TermSuggestionBuilder.java]: @cbuescher fixed

**Comment by abeyad** [comment]
@cbuescher I changed all `Locale.US` to `Locale.ROOT`

File: core/src/test/java/org/elasticsearch/index/query/HasChildQueryParserTests.java [hunk 4]
```diff
@@ -26,23 +26,23 @@
 
 public class HasChildQueryParserTests extends ESTestCase {
     public void testMinFromString() {
-        assertThat("fromString(min) != MIN", ScoreMode.Min, equalTo(HasChildQueryParser.parseScoreMode("min")));
+        assertThat("resolve(min) != MIN", ScoreMode.Min, equalTo(HasChildQueryParser.parseScoreMode("min")));
```
Comment (cbuescher) [inline core/src/test/java/org/elasticsearch/index/query/HasChildQueryParserTests.java]: Why these changes? I first thought it's because the method rename but it seems unrelated to your change now. If so, can you change this back?

File: core/src/main/java/org/elasticsearch/search/suggest/term/TermSuggestionBuilder.java [hunk 5]
```diff
@@ -493,9 +526,18 @@ public StringDistanceImpl readFrom(final StreamInput in) throws IOException {
             return values()[ordinal];
         }
 
-        public static StringDistanceImpl fromString(final String str) {
+        public static StringDistanceImpl resolve(final String str) {
             Objects.requireNonNull(str, "Input string is null");
-            return valueOf(str.toUpperCase(Locale.ROOT));
+            try {
+                return valueOf(str.toUpperCase(Locale.ROOT));
+            } catch (IllegalArgumentException e) {
```
Comment (cbuescher) [inline core/src/main/java/org/elasticsearch/search/suggest/term/TermSuggestionBuilder.java]: I don't think raising en exception to save a few lines of code here is a good idea, please change this back to how it was before.

File: core/src/test/java/org/elasticsearch/index/query/HasChildQueryParserTests.java (hunk 4 above)
Comment (abeyad) [inline core/src/test/java/org/elasticsearch/index/query/HasChildQueryParserTests.java]: good catch, it was my IDE, my fault

File: core/src/main/java/org/elasticsearch/search/suggest/term/TermSuggestionBuilder.java (hunk 5 above)
Comment (cbuescher) [inline core/src/main/java/org/elasticsearch/search/suggest/term/TermSuggestionBuilder.java]: By that I mean a switch or something similar to what SuggestUtils#resolveDistance() does.

**Comment by abeyad** [comment]
@cbuescher made the changes

**Comment by cbuescher** [comment]
Thanks, LGTM now.
